R/grid_helpers.R

```diff
@@ -370,21 +391,74 @@ compute_grid_info <- function(workflow, grid) {
                    n = res$.num_models,
                    res$.msg_preprocessor)
 
-  res %>%
+  res <- res %>%
     dplyr::select(-.num_models) %>%
     dplyr::relocate(dplyr::starts_with(".msg"))
+
+  # ----------------------------------------------------------------------------
+  # Finally, iterate across postprocessors. Conditioning on an .iter_config,
+  # make an iterator for each postprocessing candidate set (if any).
+  if (!any_parameters_postprocessor) {
+    return(res)
+  }
+
+  res <-
+    res %>%
+    dplyr::group_nest(.iter_config, keep = TRUE) %>%
+    dplyr::mutate(
+      data = purrr::map(data, make_iter_postprocessor, parameters_postprocessor$id)
+    ) %>%
+    tidyr::unnest(cols = data) %>%
+    dplyr::relocate(dplyr::starts_with(".iter"), dplyr::starts_with(".msg"))
+
+  res
 }
 
 make_iter_config <- function(dat) {
   # Compute labels for the models *within* each preprocessing loop.
-  num_submodels <- purrr::map_int(dat$.submodels, ~ length(unlist(.x)))
+  num_submodels <- purrr::map_int(
+    dat$.submodels,
+    function(.x) {if (length(.x) == 0) 0 else length(.x[[1]])}
+  )
   num_models <- sum(num_submodels + 1) # +1 for the model being trained
   .mod_label <- recipes::names0(num_models, "Model")
   .iter_config <- paste(dat$.lab_pre[1], .mod_label, sep = "_")
   .iter_config <- vctrs::vec_chop(.iter_config, sizes = num_submodels + 1)
   tibble::tibble(.iter_config  = .iter_config)
 }
 
+make_iter_postprocessor <- function(data, post_params) {
+  nested_by_post <- "post" %in% names(data)
+  if (nested_by_post) {
+    data <- data %>% unnest(post)
+  }
+
+  data %>%
+    mutate(
+      .iter_postprocessor = seq_len(nrow(.)),
+      .msg_postprocessor = new_msgs_postprocessor(
+        i = .iter_postprocessor,
+        n = max(.iter_postprocessor),
+        msgs_model = .msg_model
+      ),
+      .iter_config_post = purrr::map2(
+        .iter_config,
+        .iter_postprocessor,
+        make_iter_config_post
+      )
+    ) %>%
+    select(-.iter_config) %>%
+    nest(post = c(any_of(post_params), ".iter_postprocessor", ".msg_postprocessor", ".iter_config_post"))
```

Comment:
We actually want postprocessor parameters names both inside and outside of `.submodels`, but this results in them only being inside of `.submodels`.